jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java

```diff
@@ -241,7 +241,14 @@ MethodHandle cloneWithNewType(MethodType newType) {
 	/*[IF ]*/
 	// Until the JIT can synthesize calls to virtual methods, we must synthesize calls to these static ones instead
 	/*[ENDIF]*/
-	private static MethodHandle asType(MethodHandle mh, MethodType newType) { return mh.asType(newType); }
+	private static MethodHandle asType(MethodHandle mh, MethodType newType) {
+		/* JIT can easily propagate type information and fold the if when it can prove early return always happen */
+		if (mh.type == newType) {
+			return mh;
+		} else {
+			return mh.asType(newType);
```

Comment:
asType startes with:
```
	public MethodHandle asType(MethodType newType) throws ClassCastException {
		if (this.type.equals(newType)) {
			return this;
		}
```
which is doing the same check.  Changing that implementation to use `==` instead of `.equals` would prevent duplicate checks.  Does that allow the jit to do the same early return?

Reply:
Yes, but since the JIT will inline the early return to allow more optimizations kick in on the path where early return happens, putting the early return in the method where the real `asType` implementation lives will make the JIT inline too much, which might confuse other optimizations.